File: Holes/imfill.cpp
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

#include <opencv2/opencv.hpp>

namespace fs = std::filesystem;

struct PipelineResult {
    cv::Mat binary;
    cv::Mat flooded_background;
    cv::Mat holes;
    cv::Mat filled;
};

cv::Mat thresholdForeground(const cv::Mat& gray, int threshold_value) {
    if (gray.empty() || gray.type() != CV_8UC1) {
        throw std::invalid_argument(
            "thresholdForeground expects a non-empty 8-bit grayscale image");
    }
    if (threshold_value < 0 || threshold_value > 255) {
        throw std::invalid_argument("threshold must be in the range [0, 255]");
    }

    cv::Mat binary;
    cv::threshold(
        gray, binary, threshold_value, 255, cv::THRESH_BINARY_INV);
    return binary;
}

PipelineResult fillHoles(const cv::Mat& binary) {
    if (binary.empty() || binary.type() != CV_8UC1) {
        throw std::invalid_argument(
            "fillHoles expects a non-empty 8-bit single-channel mask");
    }

    cv::Mat invalid;
    cv::inRange(binary, cv::Scalar(1), cv::Scalar(254), invalid);
    if (cv::countNonZero(invalid) != 0) {
        throw std::invalid_argument(
            "fillHoles expects a binary mask containing only 0 and 255");
    }

    cv::Mat padded;
    cv::copyMakeBorder(
        binary, padded, 1, 1, 1, 1, cv::BORDER_CONSTANT, cv::Scalar(0));
    cv::Mat flooded_padded = padded.clone();
    cv::Mat flood_mask =
        cv::Mat::zeros(flooded_padded.rows + 2, flooded_padded.cols + 2, CV_8U);
    cv::floodFill(flooded_padded, flood_mask, cv::Point(0, 0), cv::Scalar(255));

    cv::Mat flooded_background =
        flooded_padded(cv::Rect(1, 1, binary.cols, binary.rows)).clone();
    cv::Mat holes;
    cv::bitwise_not(flooded_background, holes);
    cv::Mat filled;
    cv::bitwise_or(binary, holes, filled);
    return {binary.clone(), flooded_background, holes, filled};
}

void writeImage(const fs::path& path, const cv::Mat& image) {
    if (!cv::imwrite(path.string(), image)) {
        throw std::runtime_error("OpenCV could not write " + path.string());
    }
}

void printUsage(const char* program) {
    std::cerr << "Usage: " << program
              << " [image] [--threshold 0..255] [--output-dir DIR] [--display]\n";
}

bool parseInteger(const std::string& value, int& parsed) {
    try {
        std::size_t consumed = 0;
        parsed = std::stoi(value, &consumed);
        return consumed == value.size();
    } catch (const std::exception&) {
        return false;
    }
}

int main(int argc, char** argv) {
    fs::path image_path = "nickel.jpg";
    fs::path output_directory = "output";
    int threshold_value = 220;
    bool display = false;
    bool positional_image_seen = false;

    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--threshold" && index + 1 < argc) {
            const std::string value = argv[++index];
            if (!parseInteger(value, threshold_value)) {
                std::cerr << "error: invalid integer for --threshold: "
                          << value << '\n';
                return 2;
            }
        } else if (argument == "--output-dir" && index + 1 < argc) {
            output_directory = argv[++index];
        } else if (argument == "--display") {
            display = true;
        } else if (argument == "--help" || argument == "-h") {
            printUsage(argv[0]);
            return 0;
        } else if (!argument.empty() && argument.front() != '-' &&
                   !positional_image_seen) {
            image_path = argument;
            positional_image_seen = true;
        } else {
            printUsage(argv[0]);
            return 2;
        }
    }

    try {
        cv::Mat gray = cv::imread(image_path.string(), cv::IMREAD_GRAYSCALE);
        if (gray.empty()) {
            throw std::runtime_error("Could not read input image: " +
                                     image_path.string());
        }

        const cv::Mat binary = thresholdForeground(gray, threshold_value);
        const PipelineResult result = fillHoles(binary);
        fs::create_directories(output_directory);
        writeImage(output_directory / "01-thresholded.png", result.binary);
        writeImage(
            output_directory / "02-flooded-background.png",
            result.flooded_background);
        writeImage(output_directory / "03-holes.png", result.holes);
        writeImage(output_directory / "04-filled.png", result.filled);

        std::cout << "input=" << image_path << '\n'
                  << "threshold=" << threshold_value << '\n'
                  << "foreground_pixels_before="
                  << cv::countNonZero(result.binary) << '\n'
                  << "filled_hole_pixels=" << cv::countNonZero(result.holes)
                  << '\n'
                  << "foreground_pixels_after="
                  << cv::countNonZero(result.filled) << '\n'
                  << "outputs=" << fs::absolute(output_directory) << '\n';

        if (display) {
            cv::imshow("Thresholded", result.binary);
            cv::imshow("Flooded exterior", result.flooded_background);
            cv::imshow("Holes", result.holes);
            cv::imshow("Filled", result.filled);
            cv::waitKey(0);
            cv::destroyAllWindows();
        }
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 2;
    }
    return 0;
}

File: Holes/imfill.py
#!/usr/bin/env python3
"""Fill enclosed holes in a binary foreground mask with OpenCV."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2
import numpy as np
from numpy.typing import NDArray

Image = NDArray[np.uint8]


def threshold_foreground(gray: Image, threshold_value: int = 220) -> Image:
    """Convert a grayscale image to a white-foreground binary mask."""
    if (
        gray is None
        or gray.size == 0
        or gray.ndim != 2
        or gray.dtype != np.uint8
    ):
        raise ValueError(
            "threshold_foreground expects a non-empty uint8 grayscale image"
        )
    if not 0 <= threshold_value <= 255:
        raise ValueError("threshold_value must be in the range [0, 255]")

    _, binary = cv2.threshold(
        gray, threshold_value, 255, cv2.THRESH_BINARY_INV
    )
    return binary


def fill_holes(binary: Image) -> tuple[Image, Image, Image]:
    """Return the filled mask, flooded background, and isolated hole mask.

    A one-pixel black border guarantees that the flood-fill seed belongs to
    the exterior background, even when foreground pixels touch the source
    image boundary.
    """
    if (
        binary is None
        or binary.size == 0
        or binary.ndim != 2
        or binary.dtype != np.uint8
    ):
        raise ValueError("fill_holes expects a non-empty uint8 single-channel mask")

    values = np.unique(binary)
    if not np.all(np.isin(values, (0, 255))):
        raise ValueError("fill_holes expects a binary mask containing only 0 and 255")

    padded = cv2.copyMakeBorder(binary, 1, 1, 1, 1, cv2.BORDER_CONSTANT, value=0)
    flooded_padded = padded.copy()
    height, width = flooded_padded.shape
    flood_mask = np.zeros((height + 2, width + 2), dtype=np.uint8)
    cv2.floodFill(flooded_padded, flood_mask, (0, 0), 255)

    flooded_background = flooded_padded[1:-1, 1:-1]
    holes = cv2.bitwise_not(flooded_background)
    filled = cv2.bitwise_or(binary, holes)
    return filled, flooded_background, holes


def run_pipeline(
    input_path: Path, threshold_value: int = 220
) -> tuple[Image, Image, Image, Image]:
    """Read an image, threshold it, and fill enclosed background regions."""
    gray = cv2.imread(str(input_path), cv2.IMREAD_GRAYSCALE)
    if gray is None:
        raise FileNotFoundError(f"Could not read input image: {input_path}")

    binary = threshold_foreground(gray, threshold_value)
    filled, flooded_background, holes = fill_holes(binary)
    return binary, flooded_background, holes, filled


def write_outputs(
    output_directory: Path,
    binary: Image,
    flooded_background: Image,
    holes: Image,
    filled: Image,
) -> None:
    """Write every stage so the headless CLI is easy to inspect and test."""
    output_directory.mkdir(parents=True, exist_ok=True)
    outputs = {
        "01-thresholded.png": binary,
        "02-flooded-background.png": flooded_background,
        "03-holes.png": holes,
        "04-filled.png": filled,
    }
    for filename, image in outputs.items():
        path = output_directory / filename
        if not cv2.imwrite(str(path), image):
            raise OSError(f"OpenCV could not write output image: {path}")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Fill holes in a thresholded foreground mask with OpenCV."
    )
    parser.add_argument(
        "image",
        nargs="?",
        type=Path,
        default=Path(__file__).with_name("nickel.jpg"),
        help="Input image (default: bundled nickel.jpg)",
    )
    parser.add_argument(
        "--threshold",
        type=int,
        default=220,
        help="Binary-inverse threshold in [0, 255] (default: 220)",
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path("output"),
        help="Directory for the four pipeline images (default: output)",
    )
    parser.add_argument(
        "--display",
        action="store_true",
        help="Open GUI windows after writing the output images",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    try:
        stages = run_pipeline(args.image, args.threshold)
        write_outputs(args.output_dir, *stages)
    except (FileNotFoundError, OSError, ValueError) as error:
        print(f"error: {error}")
        return 2

    binary, _, holes, filled = stages
    added_pixels = int(cv2.countNonZero(holes))
    print(f"input={args.image}")
    print(f"threshold={args.threshold}")
    print(f"foreground_pixels_before={int(cv2.countNonZero(binary))}")
    print(f"filled_hole_pixels={added_pixels}")
    print(f"foreground_pixels_after={int(cv2.countNonZero(filled))}")
    print(f"outputs={args.output_dir.resolve()}")

    if args.display:
        for label, image in zip(
            ("Thresholded", "Flooded exterior", "Holes", "Filled"), stages
        ):
            cv2.imshow(label, image)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
